`expose_stan_functions` fails with `..._lp` functions in rstan 2.26

#### Summary:
I am getting an error on different systems (Mac/Linux) when trying to use `expose_stan_functions` with an `..._lp` function.

#### Description:
A simple example is pasted below. The code works with version 2.21.8 but not 2.26.23.

#### Reproducible Steps:
``` r
library("rstan")
#> Loading required package: StanHeaders
#> 
#> rstan version 2.26.23 (Stan version 2.26.1)
#> For execution on a local, multicore CPU with excess RAM we recommend calling
#> options(mc.cores = parallel::detectCores()).
#> To avoid recompilation of unchanged Stan programs, we recommend calling
#> rstan_options(auto_write = TRUE)
#> For within-chain threading using `reduce_sum()` or `map_rect()` Stan functions,
#> change `threads_per_chain` option:
#> rstan_options(threads_per_chain = 1)

## make temporary directory
tmpdir <- tempdir()

## definfe test stan function
stan_function <- 'functions{ void test() {} }'
stan_function_lp <- 'functions{ void test_lp() {} }'

## write stan function to file
writeLines(stan_function, file.path(tmpdir, "test.stan"))
writeLines(stan_function_lp, file.path(tmpdir, "test_lp.stan"))

## compile stan function
expose_stan_functions(file.path(tmpdir, "test.stan"))
message("Success")
#> Success

expose_stan_functions(file.path(tmpdir, "test_lp.stan"))
#> 
#> WARNING: The tools required to build C++ code for R were not found.
#> 
#> Please install Command Line Tools for XCode (or equivalent).
#> Error in expose_stan_functions(file.path(tmpdir, "test_lp.stan")): Compilation failed!
```

<sup>Created on 2023-09-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

#### Current Output:
An unspecific and incorrect error about the toolchain not being present. When using `show_compiler_warnings = TRUE` the error is
```
In file included from filea3573a329a26.cpp:9:
In file included from /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/rstan/include/exporter.h:4:
In file included from /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include/RcppCommon.h:169:
In file included from /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include/Rcpp/as.h:25:
/Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library/Rcpp/include/Rcpp/internal/Exporter.h:31:28: error: no matching constructor for initialization of 'stan::math::accumulator<double>'
                    Exporter( SEXP x ) : t(x){}
 ```

#### Expected Output:
No error.

#### RStan Version:
2.26.23

#### R Version:
R version 4.3.1 (2023-06-16)

#### Operating System:
Tried OS X 13.5.1 and Ubuntu 22.04.3 LTS

--- a/rstan/rstan/R/expose_stan_functions.R
+++ b/rstan/rstan/R/expose_stan_functions.R
@@ -30,9 +30,9 @@ expose_stan_functions_hacks <- function(code, includes = NULL) {
                 code, sep = "\n")
   code <- gsub("// [[stan::function]]",
                "// [[Rcpp::export]]", code, fixed = TRUE)
-  code <- gsub("stan::math::accumulator<double>& lp_accum__, std::ostream* pstream__ = nullptr){",
+  code <- gsub("stan::math::accumulator<double>& lp_accum__,(\\n)?(\\s*)?std::ostream\\* pstream__ = (nullptr|0))(\\s*)?\\{",
                "std::ostream* pstream__ = nullptr){\nstan::math::accumulator<double> lp_accum__;",
-               code, fixed = TRUE)
+               code)
   code <- gsub("pstream__(\\s*|)=(\\s*|)nullptr", "pstream__ = 0", code)
   return(code)
 }
@@ -50,8 +50,8 @@ expose_stan_functions <- function(stanmodel, includes = NULL,
     mc <- get_stancode(stanmodel)
   }
   else if(is.character(stanmodel)) {
-    if(length(stanmodel) == 1) mc <- get_model_strcode(stanmodel, NULL)
-    else mc <- get_model_strcode(model_code = stanmodel)
+    if(length(stanmodel) == 1) mc <- rstan:::get_model_strcode(stanmodel, NULL)
+    else mc <-  rstan:::get_model_strcode(model_code = stanmodel)
   }
   else stop("'stanmodel' is not a valid object")
 
@@ -63,7 +63,7 @@ expose_stan_functions <- function(stanmodel, includes = NULL,
   r <- stanc(model_code = mc, model_name = "User-defined functions",
              allow_undefined = TRUE,
              standalone_functions = TRUE)
-  code <- expose_stan_functions_hacks(r$cppcode, includes)
+  code <- rstan:::expose_stan_functions_hacks(r$cppcode, includes)
 
   WINDOWS <- .Platform$OS.type == "windows"
   R_version <- with(R.version, paste(major, minor, sep = "."))

--- a/rstan/rstan/R/expose_stan_functions.R
+++ b/rstan/rstan/R/expose_stan_functions.R
@@ -51,7 +51,7 @@ expose_stan_functions <- function(stanmodel, includes = NULL,
   }
   else if(is.character(stanmodel)) {
     if(length(stanmodel) == 1) mc <- get_model_strcode(stanmodel, NULL)
-    else mc <-  get_model_strcode(model_code = stanmodel)
+    else mc <- get_model_strcode(model_code = stanmodel)
   }
   else stop("'stanmodel' is not a valid object")
 

--- a/rstan/rstan/R/expose_stan_functions.R
+++ b/rstan/rstan/R/expose_stan_functions.R
@@ -50,8 +50,8 @@ expose_stan_functions <- function(stanmodel, includes = NULL,
     mc <- get_stancode(stanmodel)
   }
   else if(is.character(stanmodel)) {
-    if(length(stanmodel) == 1) mc <- rstan:::get_model_strcode(stanmodel, NULL)
-    else mc <-  rstan:::get_model_strcode(model_code = stanmodel)
+    if(length(stanmodel) == 1) mc <- get_model_strcode(stanmodel, NULL)
+    else mc <-  get_model_strcode(model_code = stanmodel)
   }
   else stop("'stanmodel' is not a valid object")
 
@@ -63,7 +63,7 @@ expose_stan_functions <- function(stanmodel, includes = NULL,
   r <- stanc(model_code = mc, model_name = "User-defined functions",
              allow_undefined = TRUE,
              standalone_functions = TRUE)
-  code <- rstan:::expose_stan_functions_hacks(r$cppcode, includes)
+  code <- expose_stan_functions_hacks(r$cppcode, includes)
 
   WINDOWS <- .Platform$OS.type == "windows"
   R_version <- with(R.version, paste(major, minor, sep = "."))

